lib/Sema/CSApply.cpp

```diff
@@ -955,6 +955,155 @@ namespace {
+      /// parameter, the callee type is the uncurried type of the callee
+      /// expression, i.e., '(args) -> result' vs. '(self) -> (args) -> result'.
+      if (baseExpr) {
+        calleeFnTy = calleeFnTy->getResult()->castTo<FunctionType>();
+      }
+
+      const auto &appliedPropertyWrappers =
+          solution.appliedPropertyWrappers[locator.getAnchor()];
+      const auto calleeDeclRef = resolveConcreteDeclRef(
+          dyn_cast<AbstractFunctionDecl>(declOrClosure), locator);
+
+      auto *const calleeParamList = getParameterList(declOrClosure);
+      const auto calleeParams = calleeFnTy->getParams();
+
+      // Rebuild the callee params: SILGen knows how to emit property-wrapped
+      // parameters, but the corresponding parameter types need to match the
+      // backing wrapper types.
+      SmallVector<AnyFunctionType::Param, 4> newCalleeParams;
+      newCalleeParams.reserve(calleeParams.size());
+
+      // Pass all the thunk parameters to the call.
+      SmallVector<Argument, 4> args;
+      unsigned appliedWrapperIndex = 0;
+      for (const auto idx : indices(*thunkParamList)) {
+        auto *const thunkParamDecl = thunkParamList->get(idx);
+
+        const auto calleeParam = calleeParams[idx];
+        const auto calleeParamType = calleeParam.getParameterType();
+        const auto thunkParamType = thunkParamDecl->getType();
+
+        Expr *paramRef = new (ctx)
+            DeclRefExpr(thunkParamDecl, DeclNameLoc(), /*implicit*/ true);
+        paramRef->setType(thunkParamDecl->isInOut()
+                              ? LValueType::get(thunkParamType)
+                              : thunkParamType);
+        cs.cacheType(paramRef);
+
+        paramRef = coerceToType(paramRef,
+                                thunkParamDecl->isInOut()
+                                    ? LValueType::get(calleeParamType)
+                                    : calleeParamType,
+                                locator);
+
+        auto *const calleeParamDecl = calleeParamList->get(idx);
+        if (calleeParamDecl->hasAttachedPropertyWrapper()) {
+          // Rewrite the parameter ref to the backing wrapper initialization
+          // expression.
+          auto &appliedWrapper = appliedPropertyWrappers[appliedWrapperIndex++];
+
+          using ValueKind = AppliedPropertyWrapperExpr::ValueKind;
+          ValueKind valueKind = (appliedWrapper.initKind ==
+                                         PropertyWrapperInitKind::ProjectedValue
+                                     ? ValueKind::ProjectedValue
+                                     : ValueKind::WrappedValue);
+
+          paramRef = AppliedPropertyWrapperExpr::create(
+              ctx, calleeDeclRef, calleeParamDecl, SourceLoc(),
+              appliedWrapper.wrapperType, paramRef, valueKind);
+          cs.cacheExprTypes(paramRef);
+
+          newCalleeParams.push_back(calleeParam.withType(paramRef->getType()));
+
+          // FIXME: inout/vararg?
```

Comment:
I see, will mark `inout` as a `TODO:` then! Was it supposed to be an error to have an `inout` wrapped parameter?

> Good point about variadics, presumably this should work and it doesn't today

Interesting, I was expecting `@Wrapper value: Int...` to mean `[Wrapper<Int>]` rather than `Wrapper<[Int]>`.

Reply:
I think a variadic wrapped parameter would have to mean the wrapped value is an array type because of the restriction that the parameter to `init(wrappedValue:)` must either exactly match the `wrappedValue` type or be an `autoclosure` returning the `wrappedValue` type.

> Was it supposed to be an error to have an inout wrapped parameter?

Oops, yeah I think that should be an error at the declaration. Hopefully no code is actually using an `inout` wrapped parameter, because the synthesized accessors inside the body don't actually let you mutate the parameter anyway. I also suspect the compiler will crash in the case where the wrapper has `init(projectedValue:)`

Reply:
Oh, I misread your comment about variadics as `Wrapper<Int>`, I understand what you're saying now. Interesting! My brain definitely interpreted `@Wrapper value: Int...` as a single wrapper.